proxy/client.go

```diff
@@ -0,0 +1,233 @@
+
+const (
+	// timeout for upstream /sync requests (after which it will send back
+	// an empty response)
+	syncTimeout = 60 * time.Second
+)
+
+type MatrixClient struct {
+	AccessToken string
+
+	url string
+
+	NextSyncBatch string
+
+	// our client for the upstream connection
+	httpClient http.Client
+}
+
+func NewClient(url string, accessToken string) *MatrixClient {
+	if !strings.HasSuffix(url, "/") {
+		url += "/"
+	}
+	return &MatrixClient{url: url, AccessToken: accessToken}
+}
+
+// an error returned when a matrix endpoint returns a non-200 whose body is
+// not a matrix error
+type HttpError struct {
+	StatusCode  int
+	ContentType string
+	Body        []byte
+}
+
+func (s *HttpError) Error() string {
+	return string(s.Body)
+}
+
+// an error returned when a matrix endpoint returns a standard matrix error
+type MatrixError struct {
+	HttpError
+	Details MatrixErrorDetails
+}
+
+func (e *MatrixError) Error() string {
+	return fmt.Sprintf("%s (%s)", e.Details.ErrCode, e.Details.Error)
+}
+
+type MatrixErrorDetails struct {
+	ErrCode string `json:"errcode"`
+	Error   string `json:"error"`
+}
+
+// Sync sends the sync request, and returns the body of the response,
+// or an error.
+//
+// It keeps track of the 'next_batch' from the result, and uses it to se the
+// 'since' parameter for the next call.
+//
+// If 'waitForEvents' is set, we will set a non-zero timeout value so that the
+// HTTP request blocks until events are ready to be read. Otherwise the call
+// will return immediately, even if there are no new events.
+//
+// Note that this method is not thread-safe; there should be only one concurrent
+// call per MatrixClient.
+//
+// If /sync returns a non-200 response, the error returned will be a MatrixError.
+func (s *MatrixClient) Sync(waitForEvents bool) ([]byte, error) {
+	timeout := syncTimeout
+	if !waitForEvents {
+		timeout = 0
+	}
+	params := url.Values{
+		"access_token": {s.AccessToken},
+		"timeout":      {fmt.Sprintf("%d", timeout/time.Millisecond)},
+	}
+	if s.NextSyncBatch != "" {
+		params.Set("since", s.NextSyncBatch)
+	}
+
+	body, err := s.get("_matrix/client/r0/sync", params)
+	if err != nil {
+		return nil, err
+	}
+
+	// we need the 'next_batch' token, so fish that out
+	next_batch, err := extractNextBatch(body)
+	if err != nil {
+		return nil, err
+	}
+	log.Println("Got next_batch:", next_batch)
+	s.NextSyncBatch = next_batch
+
+	return body, nil
+}
+
+// extractNextBatch fishes the 'next_batch' member out of the JSON response from
+// /sync.
+func extractNextBatch(httpBody []byte) (string, error) {
+	type syncResponse struct {
+		NextBatch string `json:"next_batch"`
+	}
+	var sr syncResponse
+	if err := json.Unmarshal(httpBody, &sr); err != nil {
+		return "", err
+	}
+
+	if sr.NextBatch == "" {
+		return "", fmt.Errorf("/sync response missing next_batch")
+	}
+
+	return sr.NextBatch, nil
+}
+
+func (s *MatrixClient) SendMessage(roomID string, eventType string,
+	txnID string, content []byte) (string, error) {
```

Comment:
I'm a little surprised that txnID is a parameter on the public interface, rather than something you generate in MatrixClient itself; surely you shouldn't be de-duplicating API calls?